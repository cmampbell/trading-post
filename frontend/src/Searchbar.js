import React, { useEffect, useState } from "react";
import { TextField, Autocomplete } from "@mui/material";

const Searchbar = ({ searchInput, setSearchInput, cards, setSelectedCard, selectedCard, isLoading }) => {
    const handleInputChange = (evt, newValue) => {
        setSearchInput(()=> newValue)
    }

    const handleValueChange = (evt, newValue) => {
        setSelectedCard(()=> newValue)
    }

    useEffect(()=> {
        return () => {
            setSearchInput('')
        setSelectedCard(()=> '')
        }
    }, [])

    return (
        <>
        <Autocomplete
            inputValue={searchInput}
            onInputChange={handleInputChange}
            value={selectedCard}
            onChange={handleValueChange}
            id="searchbar"
            freeSolo
            selectOnFocus
            handleHomeEndKeys
            clearOnBlur
            clearOnEscape
            autoSelect
            loading={isLoading}
            loadingText='Loading...'
            noOptionsText='No options found.'
            options={cards.map((card) => ({id: card.oracle_id, label: card.name}))}
            renderInput={(params) => <TextField {...params} label="Card Name" />}
            sx={{width: '90%'}}
        />
        </>
    )
}

export default Searchbar;