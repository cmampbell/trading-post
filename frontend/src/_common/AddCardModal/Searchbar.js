import React, { useEffect, useState } from "react";
import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import Button from "@mui/material/Button";
import Modal from "@mui/material/Modal";
import Box from "@mui/material/Box";
import IconButton from "@mui/material/IconButton";
import Grid  from "@mui/material/Grid";
import PhotoCameraIcon from '@mui/icons-material/PhotoCamera';
import CardService from "../../Api/CardService"
import WebcamCardReader from "./WebcamCardReader";

/* Returns MUI autocomplete component https://mui.com/material-ui/react-autocomplete/
*
* User begins typing, after half a second after last input change, API call is made
*   matching card names autofill a dropdown menu, user can pick
*   selected card state from AddCardModal is set based on user selection
*       Searchbar cleared
*       selected card is populated into CardDetailsBox on AddCardModal
*
* User can also choose to scan card with their camera. The user captures a photo of
* the card, and the Webcam component calls getCardWithCamera, with the tesseract
* result of what it thinks the card name is. We then set the selectedCard for the
* AddCardModal equal to the first entry of the API results.
*
*   The first entry will be the correct card if tesseract was able to extract it,
*   or the first returned card from the name like search.
*
*/
const Searchbar = ({ setSelectedCard, selectedCard }) => {
    const [searchInput, setSearchInput] = useState('');
    const [cardOptions, setCardOptions] = useState([])
    const [isLoading, setIsLoading] = useState(false);
    const [cameraOpen, setCameraOpen] = useState(false)

    useEffect(() => {
        let timerID;
        if (searchInput && !selectedCard) {
            setIsLoading(true)
            timerID = setTimeout(async () => {
                try {
                    const cards = await CardService.getCardsByName(searchInput)
                    console.log(cards);
                    setCardOptions(() => [...cards])
                    setIsLoading(() => false)
                } catch (err) {
                    console.log(err)
                }
            }, 500);
        }
        return () => {
            // clear ID if search input changes bc user is typing
            clearTimeout(timerID)
            // Set cards to empty array to clear old results
            setCardOptions(() => [])
        }
    }, [searchInput, selectedCard])

    const handleInputChange = (evt, newValue) => {
        // controlled autocomplete
        setSearchInput(() => newValue)
    }

    const handleValueChange = (evt, newValue) => {
        // value changes when user makes a selection
        if (newValue) setSelectedCard(() => newValue)
    }

    const openCameraModal = () => {
        setCameraOpen(() => true)
    }

    const closeCameraModal = () => {
        setCameraOpen(() => false)
    }

    const getCardWithCamera = async (cardNameGuess) => {
        if (cardNameGuess) setSearchInput(() => cardNameGuess);
        try {
            const card = await CardService.getCardsByName(cardNameGuess);
            setSelectedCard(() => ({ name: card[0].name, id: card[0].oracle_id }))
        } catch (err) {
            console.error(err)
        }
    }

    return (
        <>
            <Grid container sx={{width: '100%'}}>
                <Grid item xs={10}>
                    <Autocomplete
                        inputValue={searchInput}
                        onInputChange={handleInputChange}
                        value={selectedCard}
                        onChange={handleValueChange}
                        id="searchbar"
                        freeSolo
                        selectOnFocus
                        handleHomeEndKeys
                        clearOnEscape
                        autoSelect
                        loading={isLoading}
                        loadingText='Loading...'
                        noOptionsText='No cards found.'
                        options={cardOptions.map((card) => ({ id: card.oracle_id, label: card.name }))}
                        renderInput={(params) => <TextField
                            {...params}
                            label="Card Name"
                        />}
                        sx={{ marginLeft: '5%', width: '90%' }}
                        filterOptions={(x) => x}
                    />
                </Grid>
                <Grid item xs={1}>
                    <IconButton onClick={openCameraModal} color="primary" >
                        <PhotoCameraIcon sx={{ fontSize: '35px' }}/>
                    </IconButton>
                </Grid>
            </Grid>

            <Modal
                open={cameraOpen}
                onClose={closeCameraModal}
                aria-labelledby="camera-modal-title"
                aria-describedby="camera-modal"
            >
                <Box
                    sx={{
                        display: 'flex',
                        justifyContent: 'flex-start',
                        flexDirection: 'column',
                        alignItems: 'center',
                        marginTop: '8vh',
                        height: '85vh',
                        position: 'relative'

                    }}>
                    <WebcamCardReader
                        setSearchInput={setSearchInput}
                        closeCameraModal={closeCameraModal}
                        getCardWithCamera={getCardWithCamera}
                    />
                    <Button></Button>
                </Box>
            </Modal>
        </>
    )
}

export default Searchbar;