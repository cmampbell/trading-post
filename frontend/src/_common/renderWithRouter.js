import React, { isValidElement } from "react";
import { render } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";


// borrowed from: https://webup.org/blog/how-to-avoid-mocking-in-react-router-v6-tests/

export default renderWithRouter = (children, routes=[]) => {
    // options creating routes for valid children
    // else keep children as they are
    // first argument is a full route

    const options = isValidElement(children)
            ? { element: children, path: '/'}
            : children;

    // spread out the options here
    // the options are a route object
    // then routes are spread and passed after
    const router = createMemoryRouter([{...options}, ...routes],
        {
        initialEntries: [options.path],
        initialIndex: 1
    }
    );

    return render(<RouterProvider router={router}/>);
}