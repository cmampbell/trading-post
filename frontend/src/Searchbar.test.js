import React from "react";
import renderWithRouter from "./renderWithRouter";
import Searchbar from "./Searchbar";
import userEvent from '@testing-library/user-event'
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import axios from "axios";

jest.mock('axios')

const testCards = [
    { name: 'Island', image_uri: 'test_uri', usd_price: '7.65', id: '1' , setCode: 'IXL'},
    { name: 'Isuldor', image_uri: 'test_uri2', usd_price: '7.65', id: '2' , setCode: 'IXL'}
]

describe('Searchbar tests', () => {
    it('should render without crashing', () => {
        renderWithRouter(<Searchbar setSelectedCard={()=> null} selectedCard={''}/>);
    })

    it('should match snapshot', () => {
        const { asFragment } = renderWithRouter(<Searchbar setSelectedCard={()=> null} selectedCard={''}/>);

        expect(asFragment()).toMatchSnapshot();
    })

    it('should show text input', async () => {
        const { queryByLabelText } = renderWithRouter(<Searchbar setSelectedCard={()=> null} selectedCard={''}/>);

        expect(queryByLabelText('Card Name')).toBeInTheDocument();
    })

    it('should list returned cards after user types, clear cards after search term is cleared', async () => {
        axios.mockResolvedValue({data:{cards: testCards}});
        const { queryByText, queryByLabelText } = renderWithRouter(<Searchbar setSelectedCard={()=> null} selectedCard={''}/>);

        const textInput = queryByLabelText('Card Name')
        expect(textInput).toBeInTheDocument();

        await act(async () => {
            userEvent.type(textInput, 'is');
        })

        expect(textInput).toHaveValue('is')

        await waitFor(() => {
            expect(queryByText("Island")).toBeInTheDocument()
            expect(queryByText("Isuldor")).toBeInTheDocument()
        })

        axios.mockResolvedValue({data:{cards: testCards}});
        await act(async () => {
            userEvent.type(textInput, 'l');
        })

        expect(textInput).toHaveValue('isl')
        await waitFor(() => {
            expect(queryByText("Island")).toBeInTheDocument()
            expect(queryByText("Isuldor")).not.toBeInTheDocument()
        })

        await act(async ()=> {
            userEvent.clear(textInput)
        })

        await waitFor(() => {
            expect(queryByText("Island")).not.toBeInTheDocument()
        })        
    })

    it('should return no cards if no cards found', async () => {
        axios.mockResolvedValue({data:{cards: []}});
        const { queryByText, queryByLabelText } = renderWithRouter(<Searchbar setSelectedCard={()=> null} selectedCard={''}/>);

        const textInput = queryByLabelText('Card Name')
        expect(textInput).toBeInTheDocument();

        await act(async () => {
            userEvent.type(textInput, 'not-a-card');
        })

        expect(textInput).toHaveValue('not-a-card')

        await waitFor(() => {
            expect(queryByText("Island")).not.toBeInTheDocument()
        })
    })
})