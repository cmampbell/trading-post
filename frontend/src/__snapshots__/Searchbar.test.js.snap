// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon css-18nfw8h-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="searchbar"
        id="searchbar-label"
      >
        Card Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id="searchbar"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Card Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-9njbua-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6iv9kh-MuiSvgIcon-root"
      data-testid="PhotoCameraIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <circle
        cx="12"
        cy="12"
        r="3.2"
      />
      <path
        d="M9 2 7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
